### DISCUSSION TITLE
handling contextual state like auth

### DISCUSSION THREAD

**Discussion by joprice** [Omnistac/zedux#129] [Q&A] [answered]
I'm curious about approaches for handling state that is discovered at runtime, like a user session: It is initially undefined, but for atoms that deal with displaying user details or making use of the user session, it is effectively required. Say there are several atoms that manage data synced to a local cache, which is fetched using an api client, that in turn depends on a user session. Is `AtomProvider` the best approach here, where everything that requires session directly or indirectly takes these as atom instance parameters ?

**Comment by bowheart** [comment]
Hey @joprice. Normally, I'd say this isn't something you'd have to think about. Components lazily initialize atoms as they're mounted. You wouldn't typically render a `<UserName />` component, for example, until the user is logged in.

I'm not sure I'm following this part:

> Say there are several atoms that manage data synced to a local cache, which is fetched using an api client, that in turn depends on a user session

Some (pseudo-)code would be helpful. But I'm imagining this basic setup for now:

```ts
const userDataAtom = atom('userData', undefined) // populated later

const getUserName = ({ get }: AtomGetters) => get(userDataAtom)?.name

function UserName() {
  const userName = useAtomSelector(getUserName)

  return userName == null ? null : <span>{userName}</span>
}
```

The pain point with this setup that I think you're pointing out here:

> it is effectively required

is having to deal with undefined. There are a few ways to go about it. Using an `AtomProvider` like you said would work, though that might get very involved passing atom instances around.

If you really are guaranteeing that the data is defined when it's used in certain places, then a wrapper atom like this would be my preferred approach:

```ts
const loadedUserDataAtom = ion('loadedUserData', ({ get }) => {
  const userData = get(userDataAtom)

  if (!userData) {
    throw new Error('tried accessing user data before it was loaded')
  }

  return userData
})
```

Now consumers of `loadedUserDataAtom` will either throw the error, signaling to you that you broke your guarantee, or will know that the data is always defined:

```ts
// TS knows there's no need for `?.`:
const getUserName = ({ get }: AtomGetters) => get(loadedUserDataAtom).name

function UserName() {
  const userName = useAtomSelector(getUserName)

  // and no need for `== null` check:
  return <span>{userName}</span>
}
```

There may be more approaches, but let me know if I'm going in the wrong direction here.

**Comment by joprice** [comment]
Yes that's exactly the situation I was thinking of. With other approaches to state that don't attempt to represent the graph statically, I'd have some sort of "AuthGate" that ensures the nullability is handled at runtime, then from there on out the data is accessed through a non-nullable variable stored in context. For a single value like auth, it's easy enough to add the runtime check you suggested, but if you have four or five dependencies that should all be in a certain state before using them, it becomes cumbersome and error-prone. Even with a single runtime dependency like in the auth example above, I had issues where logging out caused display of errors in components that assumed the user was non-nullable. (This could have been something subtle with contexts and react native navigation, but it's not something I've had happen since I switched to passing a non-nullable reference downward since the types rule out that case). 

What I'd like in general is to have a sub-graph of atoms to be "unlocked" by a "key", similar to a object-capabilities model or how subsets of a graphql schema require parameters to access. I'm now starting to see that context is the way to go since this is a runtime verification. Just trying to figure out what I should be taking care of and what Zedux should handle. I like Zedux since it has most everything baked in that you normally have to piece together with multiple libraries, and especially that the data layer is testable outside of the UI, so I was attempting to see if I could fit the full runtime state into the static graph. I think I'll try out an approach where the base atoms are parameterized, and then provided via `AtomProvider`, along with some helpers to make usage simple and see how it turns out.

**Reply by bowheart** [reply]
Ok, yes, I think you're on the right track with an AuthGate. I think I'm gonna mostly reiterate what you said and say a lot of stuff you already know, just making sure we're on the same page.

## Accessing non-nullable provided values in React

As far as handling undefined in provided data directly in components goes, approaches you'd use to solve this in normal React apply to Zedux too. React architecture naturally "gates" child components and suspense is particularly amazing for reducing manual logic.

```ts
const userAtom = atom('user', () => api(fetchUser()))

function AuthGate() {
  // suspends until `fetchUser` resolves
  const userInstance = useAtomInstance(userAtom)

  return (
    <AtomProvider instance={userInstance}>
      <AuthedApp />
    </AtomProvider>
  )
}
```

Both Zedux and normal React Context have the same problem with handling undefined in consuming components.

```ts
// with React, you have to either specify a bad default, or handle undefined too:
const userContext = React.createContext(undefined as undefined | User)

function ChildComponentInAuthedApp() {
  const userFromZedux = useAtomValue(useAtomContext(userAtom, true)).data // can be undefined
  const userFromReact = useContext(userContext) // can also be undefined
}
```

I would solve both cases the same way - with a custom hook that throws an error if not provided:

```ts
const useUserFromZedux = () => {
  const user = useAtomValue(useAtomContext(userAtom, true)).data

  if (!user) throw new Error('user not provided')

  return user
}

const useUserFromReact = () => {
  const user = useContext(userContext)

  if (!user) throw new Error('user not provided')

  return user
}
```

I don't know of anything in React Native that could keep a component subtree mounted if e.g. AuthGate re-suspends or specifically unmounts the subtree when the user logs out (or token expires or whatever). But if there was, I'd definitely expect these errors to be thrown if the context value changes to undefined. That should hit an ErrorBoundary, which could be part of AuthGate's job.

## Accessing non-nullable values in atoms

None of that applies inside atoms. Since atoms can exist completely outside React, they know nothing about what's provided to component subtrees.

Anyway, you already know this. But now that I understand your situation a bit better, I think you're completely on the right track here. Passing atom instances as params are exactly the tool Zedux provides for this. I'd love to see what your helpers end up looking like.

We generally just handle `undefined`. One additional comment though (may or may not be helpful): I've actually experimented a bit with using React's new `use` util inside atoms. We are thinking about creating a new `inject` utility for Zedux that wraps `use` and also accepts atom templates for getting provided atom instances when the current atom is instantiated during render of a component that can consume them.

The cases it would have to handle are pretty complex and I'm not sure how viable it'll be to include in Zedux core. But barring that new util being released, this does work in React 19:

```ts
function injectReactContext<T>(context: Context<T>) {
  const { id } = injectSelf()

  return injectMemo(() => {
    const value = use(context)

    if (value == null) {
      throw new Error(`context was not provided during initial evaluation of atom "${id}"`)
    }

    return value
  }, [])
}

const userNameAtom = atom('userName', () => injectReactContext(userContext).name)
```

Though note that this takes a snapshot of the provided value - it can't react to the provided value changing. The `inject` util will solve for that (if we release it). It's also currently possible to make this support atom templates, though it would involve undocumented APIs that are changing in v2. I wouldn't recommend doing that now.

**Reply by joprice** [reply]
Yup, everything you explained fits with my understanding as well. I'll try that static and dynamic/runtime versions of this and see how they compare.

The injecting of react context is interesting. Reminds me of scala's DynamicVariable https://www.scala-lang.org/api/3.2.1/scala/util/DynamicVariable.html and of Session scope in DI systems, since the rest of the atoms don't have to knw that its dependency is dynamic. I think this is essentially the solution to the problem I'm imagining - run the whole graph, but for this run, replace these nodes with data from the context. Of course, context is more powerful because you have the cascading override ability. 

That also means that that atom is only usable from react, so perhaps an EcosystemProvider override would be used for each atom that should be wired up to context? Or the reverse - override with static values in test. Perhaps the idea from bunshi of only requesting an interface https://www.bunshi.org/advanced/interfaces/, and having it be provided by the react impl could be useful. `injectDynamic` or `injectInterface`, then a Provider that takes `impls={[...]}`. This also avoids the sentinel value problem of using undefined, -1, etc for things that should be globally provided but deferred.

**Reply by bowheart** [reply]
Wow, yes, you understand this perfectly. My goal for `inject` (or `injectDynamic` is good, or maybe `injectContext` or `injectAtomContext` to align with `useAtomContext`) is to feel very much like bunshi, but reusing atom templates, no new concepts like molecules/interfaces.

And yes, contextualized atoms could be overridden in the ecosystem to prevent throwing an error when used outside React. Though we'd probably also introduce an API for supplying atom context defaults. Maybe. This is all very preliminary thinking. I should start a new issue for it.

**Reply by joprice** [reply]
I just tried the straightforward "static" approach without context, where the session is stored in a nullable atom, the parent branches on the state by reading the atom with `useAtomValue`, and the children do the same, throwing in the case of it being undefined. However, when the child clears the atom (simulating a logout), an error is thrown in the child from a dependent atom ("Error while evaluating atom...") that consumes the now nulled-out session atom. It seems that the downstream atom is still subscribed to, causing it to re-evaluate. Ideally, the parent would have hidden this screen by that point and the child wouldn't see the null value. This is where I'm too unfamiliar with how this model is meant to work to understand the order of rendering when a parent and child are subscribing to the same atom and it gets mutated. Almost makes me think I need two atoms, one that is one update behind the other to allow the read side to survive until the parent hides it.

**Reply by bowheart** [reply]
Is there a reason you need to clear the atom (e.g. security concerns, or you just want to fully control atoms outside React)? Normally you'd give it `ttl: 0` to allow it to die when no longer in use. With that, when the parent unmounts the component subtree containing all usages of the atom, the atom is automatically destroyed.

If that's too simple a use case, there are a few things:

1. Basically your idea, creating an intermediary atom, selector, injector, or hook that holds onto the last non-nullish value. Here's an example we use for animations in our codebase:

```ts
import {
  AnyAtomInstance,
  AnyAtomTemplate,
  AtomParamsType,
  AtomStateType,
  ParamlessTemplate,
  useAtomValue,
  ZeduxHookConfig,
} from '@zedux/react';
import { useRef } from 'react';

/**
 * Returns the last non-nullish value of the provided atom to allow exit animations to finish
 * without blanking out the data in the view
 */
export const useLatchingAtomValue: {
  <A extends AnyAtomTemplate>(
    atom: A,
    params: AtomParamsType<A>,
    config?: ZeduxHookConfig
  ): AtomStateType<A>;
  <A extends AnyAtomTemplate>(atom: ParamlessTemplate<A>): AtomStateType<A>;
  <I extends AnyAtomInstance>(instance: I, params?: [], config?: ZeduxHookConfig): AtomStateType<I>;
} = <A extends AnyAtomTemplate>(
  atom: A | AnyAtomInstance,
  params?: AtomParamsType<A>,
  config?: ZeduxHookConfig
) => {
  const value = useAtomValue(atom as A, params as AtomParamsType<A>, config);
  const latchedValue = useRef(value);

  // we latch onto the old value if the value is now undefined or null
  if (value != null) {
    latchedValue.current = value;
  }

  return latchedValue.current;
};
```

(Hah, React compiler would complain about that now that I look at it. But that ref assignment would be fine in practice).

2. Zedux gives you a lot of control over order of events. You can use a promise or observable for `ttl` that prevents the atom from being destroyed until a parent component unmounts. Note that this doesn't destroy anything unless the atom goes stale (has no dependents). Example using the new `Promise.withResolvers()`:

```ts
const logoutNotifierAtom = atom('logoutNotifier', () => Promise.withResolvers())

const userAtom = atom('user', () => {
  const { promise } = injectAtomValue(logoutNotifierAtom)

  return api(fetchUser()).setTtl(promise)
})

function TopLevelAuthedComponent() {
  const { resolve } = useAtomValue(logoutNotifierAtom)

  useEffect(() => {
    return () => resolve()
  }, [])
}
```

Though, if you need to clear the atom's state, I would have expected a top-level component like this to be where you clear it. If it needs to be cleared by another atom or some child deeper in the tree, you can still wait for the top-level component to unmount before clearing, something like this:

```ts
const userAtom = atom('user', () => {
  const { get } = injectAtomGetters()
  const { store } = injectPromise(fetchUser, [])

  const clear = async () => {
    await get(logoutNotifierAtom).promise

    store.setState(state => ({ ...state, data: undefined }))
  }

  return api(store).setExports({ clear })
})
```

**Reply by joprice** [reply]
These are great solutions. I do need to run some actions to clear user data and close db sessions, so I want some kind of guarantee that it's done at least before the app closes. I haven't tried out ttls yet, so not sure how they behave. 

I think I'll need to play around with them to see which works as I'll likely need to retain the state for the exit animation case as well. I'll try it out tomorrow and provide any feedback.

**Reply by joprice** [reply]
I just got this working using the "placeholder" approach with a helper like this (not exactly the same as I'm actually calling this from a Fable Fsharp wrapper I wrote):

```js
const ignoreEmptyAtom  = (name, target)  =>
        atom(name, () => {
            const value = injectAtomValue(target);
            const current = injectStore(value);
            injectEffect(() => {
                if (value) {
                    current.setState(value);
                }
            }, [value]);
            const state = current.getState()
            if (!state) {
                throw new Error("Attempted to access atom before initial value was set");
            };
            return state
        });

const nullableAtom = atom("user", undefined)
const nonNull = ignoreEmptyAtom(nullableAtom)
```

**Reply by bowheart** [reply, marked as answer]
@joprice Cool! A few things:

- The `injectEffect` shouldn't be necessary there, unless you specifically want to wait a tick
- I'd always return the store, not the store's state, from the state factory. Zedux will create another, different store for this atom instance if you don't.

```ts
const ignoreEmptyAtom  = (name, target)  =>
    atom(name, () => {
        const value = injectAtomValue(target);
        const current = injectStore(value);
        if (value) {
            current.setState(value);
        }
        if (!current.getState()) {
            throw new Error("Attempted to access atom before initial value was set");
        };
        return current
    });
```

**Reply by joprice** [reply]
I actually put the setState in an effect because I thought it was necessary, but good to know. I remember seeing something in the docs about inline effects that mentioned the later notes in the suspense docs, but never resolved it in my mind of what exactly are the rules for inline effects. So essentially, you can think of the atom's body as being an effect block that's triggering based on dependencies, and `injectEffect` is only strictly necessary when the call should always be made at some future time. And the setState will be done synchronously before the end of the atom function, so not cause two "events" to be emitted from that atom downstream?

Returning the store makes more sense now that you point that out. I think i had just already dereferenced the state so just returned what I had at hand.

**Reply by bowheart** [reply]
> the setState will be done synchronously before the end of the atom function, so not cause two "events" to be emitted from that atom downstream?

Yes. In your case, dependents would have only been notified once anyway, since the state isn't changed until the effect runs, but yes, updating state inline can prevent dependents from being notified multiple times unnecessarily.

Zedux specifically prevents state updates on injected stores during atom evaluation from triggering another evaluation. You should still use an effect if updating state in an injected atom instance. Though in that case, you should also try to avoid the effect by e.g. using an export to update the upstream atom's state directly rather than waiting for the effect to run. The principle is the same as React's ["you might not need an effect" doc](https://react.dev/learn/you-might-not-need-an-effect).

> you can think of the atom's body as being an effect block that's triggering based on dependencies

Yes, you can treat atoms this way, though I'd only do this with specific types of atoms. For example, the docs treat query atoms this way, which works because query atoms are typically very focused on one goal: Fetching/updating one piece of data.

Atoms with less clear objectives or lots of injectors (that can use other injectors, etc) where every injector is making this assumption could get unruly.

In either case, effects are still useful for making sure a side effect only runs in a subset of the cases that could have triggered the current atom to reevaluate. But that usually won't apply when you're just forwarding an upstream state change on to an injected store.

**Reply by joprice** [reply]
That makes sense. I think I've muddled changes to parameters vs injections in my mind. I've been using the analogy of "subscribing" to changes in react props by using an effect to update local state when a prop changes, but parameters produce different atom instances in zedux. I think I need to write some tests with some injected values that change and see if I get a better feel for when the effect body is run and when an effect is actually needed in derived atoms. I'll probably hit more concrete cases to motivate it as I migrate more code over as well.

**Comment by bowheart** [comment]
@joprice FYI I opened #172 which fully implements scoped atoms, the new `inject` utility, and `ecosystem.withScope`. If you get a minute, I'd love to hear your thoughts on how this aligns with what you were envisioning here: https://github.com/Omnistac/zedux/discussions/129#discussioncomment-11352614

There are some pretty extensive examples in the [new tests](https://github.com/Omnistac/zedux/blob/d0731365621180eaf0ae6b4fb91c4461b4fe0b90/packages/react/test/integrations/scoped-atoms.test.tsx)

**Reply by joprice** [reply]
Nice I’ll give this a try soon and leave a comment.

**Reply by joprice** [reply]
I finally got around to trying this as I'm upgrading to v2 at the moment, but I'm having trouble working out how to use it. I get `Error: Scoped atom was used outside a scoped context. This atom needs to be used by a React component or inside `ecosystem.withScope` at pretty much every use-site. I'm confused about how to use it transitively. If one atom is used transitively by many others, does every single one have to use withScope and pass call `injectAtomInstance(inject` and `getters.withScope` or is there some way to have it propagate automatically? If you do need to wrap it, I guess scoped atoms should never "escape" some wrapping atom that exposes exports to avoid runtime errors?

Also, I can't seem to get it to work with selectors, which I use extensively. I tried withScope and I always get the above error. I've dug through the repo a bit and can't seem to figure out what I'm doing wrong, but probably a few more examples of real world usage would clear it up for me.

**Reply by bowheart** [reply]
Scope is infectious. It propagates to all dependents of an atom. [Here's a codesandbox](https://codesandbox.io/p/devbox/wz7cvy?migrateFrom=6vcjgt) that sets up some transitive deps with selectors. You shouldn't need to use `ecosystem.withScope` inside a selector itself - that selector will inherit the scope of its dependencies and should be called with all that scope itself.

When using atoms in React and when all atoms declare simple dependencies on others (e.g. via injectors or atom getters at the top level of the atom state factory), scope should "just work". But when used outside React or when an atom `get`s another atom dynamically in a callback, you will possibly need `ecosystem.withScope` - a single `inject` call in a top-level atom can infect everything.

Zedux does a few things to improve this:

- `injectCallback` automatically scopes the callback function to the scope of the injecting atom (including all scope it inherits from dependencies).
- exported functions are automatically scoped in the same way.

However, these auto-scoping features still require that the dependent atom capture all the required scope. Some examples:

```ts
const scopedAtom = atom('scoped-example', () => {
  return inject(someContextAtom).get()
})

// non-working example:
const notScopedAtom = ion('notScoped', ({ get }) => {
  return api('example').setExports({
    // the `notScoped` atom didn't declare `scopedAtom` as a dependency. So no scope is inherited and this call fails:
    getScopedValue: () => get(scopedAtom)
  })
})

// working example:
const transitivelyScopedAtom = ion('transitivelyScoped', ({ get, getNode }) => {
  // `scopedAtom` is now declared as a dependency. So `transitivelyScopedAtom` inherits all its scope:
  const scopedInstance = getNode(scopedAtom)

  return api('example').setExports({
    // now this call succeeds since this export is automatically wrapped in `transitivelyScopedAtom`'s scope:
    getScopedValue: () => get(scopedAtom) // Note: `scopedInstance.get()` will also work here after PR #256 ships
  })
})

// you can also capture scope manually:
const manuallyScopedAtom = ion('manuallyScoped', ({ get }) => {
  // use a throw-away `inject` call to capture scope we know we'll need later:
  inject(someContextAtom)

  return api('example').setExports({
    // now this auto-scoped export works, since `manuallyScopedAtom` captured all the same scope that `scopedAtom` requires:
    getScopedValue: () => get(scopedAtom)
  })
})
```

I'm still new to using scoped atoms myself, but I'm starting to regard `ecosystem.withScope` as a code smell. Nothing in atoms, selectors, or normal React code should need to use it. It's mostly for testing. Prefer capturing scope properly by declaring dependencies and using exports and `injectCallback` to automatically scope atom getter usages.

Also, `inject` is not supported directly in selectors right now. But there shouldn't be any problem with referencing scoped or transitively-scoped atoms/selectors from a selector, as long as that selector is itself called with scope. I see this working in the above codesandbox, but that's a simple example. Let me know if you can reproduce a problem there.

**Reply by joprice** [reply]
I've got things working somewhat with copious amounts of `withScope`. Almost every selector usage needed it. I use a lot of callbacks in useEffect and other non-reactive contexts, and maybe of them use shared selectors. I could refactor it to use exports of an atom or ion instance, and be disciplined, but I started wondering if this could be tracked in the type system of the atom, similar to how effect-ts tracks dependencies in layers, where all dependencies need to be provided before it can be used. I think the technique is just a union using Exclude to "peel off" off a layer, then you can call the function only when that type is never or some other sentinel. Right now, I'm a bit nervous as a the things I've moved into atom context touch almost every part of the code. It's about 40k loc, so not huge, but large enough that I'll only feeling comfortable after manually testing every feature. (And I tend to think the things users would move into context would be in that category of relative global concerns like auth).

**Reply by joprice** [reply]
I can't seem to resolve scoped atoms when using `injectAtomValue`. Adding withScope and injecting the scope works though. Is scope supposed to be propagated there as well?

**Reply by bowheart** [reply]
`injectAtomValue` is just an alias for `injectEcosystem().get`. And it shouldn't have a problem propagating scope provided to the current atom. The codesandbox I linked before apparently dropped some changes. I've re-applied those and added another layer using `injectAtomValue`. Seems to be working (and hopefully the codesandbox retains these latest changes this time:) https://codesandbox.io/p/devbox/wz7cvy?migrateFrom=6vcjgt

**Reply by joprice** [reply]
Weird. My selectors currently always fail unless I add a withScope. I ended up writing a helper that calls inject and adds it to get around it for now. I'll try to narrow down what I'm doing wrong after I get the refactor working end to end.

**Reply by joprice** [reply]
I was able to narrow this down to two use cases (so far) where i have to provide scope: `injectEffect`

```ts
const effectAtom = atom("effectAtom", () => {
  const user = inject(userAtom);
  const getters = injectEcosystem();
  injectEffect(() => {
    getters.withScope([user], () =>
      getters.getNodeOnce(userDetailsAtom).exports.update("Test 2"),
    );
  }, []);
});
```

which I found can be fixed by calling `injectEffect(injectCallback(...))`, but then I'm not sure if I need to provide deps twice in that case, once to the injectCallback, and once to the injectEffect.

And the other is inline selectors in `injectAtomValue`. I copied the code to a repro repo because codesandbox was throwing vm errors for me:

https://github.com/joprice/zedux-repro/compare/v2#diff-26ad4b834941d9b19ebf9db8082bd202aaf72ea0ddea85f5a8a0cb3c729cc6f2R57

**Reply by bowheart** [reply]
I'm not sure why `injectAtomValue` with inline selectors wouldn't just work here. I'll look into that.

`injectEffect` and `useEffect` are places where `withScope` will be needed. I still regard them as a code-smell in general. In your example, it would be more correct to create a real dependency on `userDetailsAtom`:

```ts
const effectAtom = atom("effectAtom", () => {
  const userDetailsInstance = injectAtomInstance(userDetailsAtom) // no `withScope` needed

  injectEffect(() => {
    userDetailsInstance.exports.update("Test 2"),
  }, [userDetailsInstance]);
});
```

This is better in general since it prevents the `userDetailsAtom` instance from being destroyed while the `effectAtom` still needs it. But there are valid cases where you need to avoid this (e.g. to prevent circular dependencies). In those cases, a `withScope` will be required.

There is one thing we can do, though it still wouldn't make your example work. `injectEffect` automatically wraps its callback in an `untrack` call and an `ecosystem.batch` call - which can only batch state updates that happen synchronously. In that same vein, we could also wrap it in `withScope` to auto-scope any operations that happen synchronously. However, that still requires that the downstream atom manually capture the scope needed by the upstream atom. It's essentially no better and more implicit. It would also create another disparity between `useEffect` and `injectEffect`. So I think I'm against this.
